common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java

```diff
@@ -379,7 +382,8 @@ public MicrosoftStsTokenRequest createTokenRequest(@NonNull final MicrosoftStsAu
             tokenRequest.setGrantType(TokenRequest.GrantTypes.AUTHORIZATION_CODE);
         }
 
-        if (PopAuthenticationSchemeInternal.SCHEME_POP.equals(authScheme.getName())) {
+        this.setAuthenticationScheme(authScheme);
```

Comment:
We shouldn't be using `setters` unless extremely necessary. This is going to make our code prone to state issues. If a `MicrosoftStsOauth2Strategy` object is modified by another piece of code that sets its authentication scheme object then we're going to run into problems.

I know the usage of setters is problem with the existing code. We have been moving away from this model and towards classes that contain `final` variables and provides `Builders` to construct the object to effectively become immutable. We should avoid the use of setters for any new code we write.

Reply:
I know, it's not ideal, but there is no other easy way to get the kid/thumbprint value other than accessing it from the authscheme. Otherwise, we have to parse the accessToken jwt and fetch the kid value from inside it, which we don't want to do.
To make the authscheme available at the time of saving token to the cache this was the only easier way without requiring a big breaking change.

Reply:
Actually, getting it from the response is not a bad idea.

We're already deserializing TokenResponse, right?

Reply:
I am checking with Brian to see the reason behind current implementation.
For now, in latest iteration I have made the authscheme part of the strategy Parameters, and removed the local variable

Reply:
We don't want to crack open AT, as this won't work for new 1P services now that token encryption is rolling out.
https://review.learn.microsoft.com/en-us/identity/microsoft-identity-platform/token-encryption?branch=main